Go crash when restoring a schedule backup

**What steps did you take and what happened:**
When attempting to restore a backup created in an Azure AKS kubernetes cluster, utilizing restic for disk backups, I received the below panic log, using v0.10.0.  The restore then becomes stuck "InProgress" and will not resume nor clean itself up when the ark pod recreates.

**What did you expect to happen:**

I expected the data to be restored successfully.

**The output of the following commands will help us better understand what's going on**:
(Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)

* `kubectl logs deployment/ark -n heptio-ark`

```
time="2019-02-22T19:57:08Z" level=info msg="Executing item action for clusterrolebindings.rbac.authorization.k8s.io" backup=cortex-ark-total-daily-2019022219314[8419/98282]
"pkg/restore/restore.go:764" restore=heptio-ark/totality
time="2019-02-22T19:57:08Z" level=info msg="Restoring ClusterRoleBinding: system:controller:disruption-controller" backup=cortex-ark-total-daily-20190222193143 logSource="p
kg/restore/restore.go:796" restore=heptio-ark/totality
time="2019-02-22T19:57:08Z" level=info msg="Executing item action for clusterrolebindings.rbac.authorization.k8s.io" backup=cortex-ark-total-daily-20190222193143 logSource=
"pkg/restore/restore.go:764" restore=heptio-ark/totality
time="2019-02-22T19:57:08Z" level=info msg="Restoring ClusterRoleBinding: system:controller:endpoint-controller" backup=cortex-ark-total-daily-20190222193143 logSource="pkg
/restore/restore.go:796" restore=heptio-ark/totality
time="2019-02-22T19:57:08Z" level=info msg="Executing item action for clusterrolebindings.rbac.authorization.k8s.io" backup=cortex-ark-total-daily-20190222193143 logSource=
"pkg/restore/restore.go:764" restore=heptio-ark/totality
time="2019-02-22T19:57:08Z" level=info msg="Restoring ClusterRoleBinding: system:controller:expand-controller" backup=cortex-ark-total-daily-20190222193143 logSource="pkg/r
estore/restore.go:796" restore=heptio-ark/totality
time="2019-02-22T19:57:08Z" level=info msg="Executing item action for clusterrolebindings.rbac.authorization.k8s.io" backup=cortex-ark-total-daily-20190222193143 logSource=
"pkg/restore/restore.go:764" restore=heptio-ark/totality
time="2019-02-22T19:57:08Z" level=info msg="Restoring ClusterRoleBinding: system:controller:generic-garbage-collector" backup=cortex-ark-total-daily-20190222193143 logSourc
e="pkg/restore/restore.go:796" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Executing item action for clusterrolebindings.rbac.authorization.k8s.io" backup=cortex-ark-total-daily-20190222193143 logSource=
"pkg/restore/restore.go:764" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Restoring ClusterRoleBinding: system:controller:horizontal-pod-autoscaler" backup=cortex-ark-total-daily-20190222193143 logSourc
e="pkg/restore/restore.go:796" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Executing item action for clusterrolebindings.rbac.authorization.k8s.io" backup=cortex-ark-total-daily-20190222193143 logSource=
"pkg/restore/restore.go:764" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Restoring ClusterRoleBinding: system:controller:job-controller" backup=cortex-ark-total-daily-20190222193143 logSource="pkg/rest
ore/restore.go:796" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Executing item action for clusterrolebindings.rbac.authorization.k8s.io" backup=cortex-ark-total-daily-20190222193143 logSource=
"pkg/restore/restore.go:764" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Restoring ClusterRoleBinding: system:controller:namespace-controller" backup=cortex-ark-total-daily-20190222193143 logSource="pk
g/restore/restore.go:796" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Executing item action for clusterrolebindings.rbac.authorization.k8s.io" backup=cortex-ark-total-daily-20190222193143 logSource=
"pkg/restore/restore.go:764" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Restoring ClusterRoleBinding: system:controller:node-controller" backup=cortex-ark-total-daily-20190222193143 logSource="pkg/res
tore/restore.go:796" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Executing item action for clusterrolebindings.rbac.authorization.k8s.io" backup=cortex-ark-total-daily-20190222193143 logSource=
"pkg/restore/restore.go:764" restore=heptio-ark/totality
time="2019-02-22T19:57:09Z" level=info msg="Restoring ClusterRoleBinding: system:controller:persistent-volume-binder" backup=cortex-ark-total-daily-20190222193143 logSource
="pkg/restore/restore.go:796" restore=heptio-ark/totality
E0222 19:57:09.608771       1 runtime.go:66] Observed a panic: "send on closed channel" (send on closed channel)
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:522
/usr/local/go/src/runtime/panic.go:513
/usr/local/go/src/runtime/chan.go:185
/usr/local/go/src/runtime/chan.go:125
/go/src/github.com/heptio/ark/pkg/restic/repository_ensurer.go:69
/go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/controller.go:202
/go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/shared_informer.go:552
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:203
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:203
/go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546
/go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/shared_informer.go:390
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71
/usr/local/go/src/runtime/asm_amd64.s:1333
panic: send on closed channel [recovered]
        panic: send on closed channel

goroutine 265 [running]:
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x108
panic(0x19f7580, 0x1f22260)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/heptio/ark/pkg/restic.newRepositoryEnsurer.func1(0x1ca7c80, 0xc0007ee780, 0x1ca7c80, 0xc000ae7380)
        /go/src/github.com/heptio/ark/pkg/restic/repository_ensurer.go:69 +0x2dc
github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(0x0, 0xc0006d9fe0, 0x0, 0x1ca7c80, 0xc0007ee780, 0x1ca7c80, 0xc000ae7380)
        /go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/controller.go:202 +0x5d
github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0xc00040bf30, 0x3, 0x3)
        /go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/shared_informer.go:552 +0x18b
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0xc000795e18, 0x42b002, 0xc0000d02b
0)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:203 +0x9c
github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
        /go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548 +0x89
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00040bf68)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000795f68, 0xdf8475800, 0x0, 0x19a0901, 0xc0004a4c60)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbe
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc00040bf68, 0xdf8475800, 0xc0004a4c60)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc0006ddd80)
        /go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546 +0x8d
github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache.(*processorListener).run-fm()
        /go/src/github.com/heptio/ark/vendor/k8s.io/client-go/tools/cache/shared_informer.go:390 +0x2a
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0003a96b0, 0xc0003223c0)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
created by github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
```


* `ark backup describe <backupname>` or `kubectl get backup/<backupname> -n heptio-ark -o yaml`
```
apiVersion: ark.heptio.com/v1
kind: Backup
metadata:
  creationTimestamp: "2019-02-22T19:51:04Z"
  generation: 1
  labels:
    app: ark
    ark-schedule: cortex-ark-total-daily
    ark.heptio.com/storage-location: default
    chart: ark-3.0.0-cs
    heritage: Tiller
    release: cortex
  name: cortex-ark-total-daily-20190222193143
  namespace: heptio-ark
  resourceVersion: "3430"
  selfLink: /apis/ark.heptio.com/v1/namespaces/heptio-ark/backups/cortex-ark-total-daily-20190222193143
  uid: 305114dc-36db-11e9-96a5-1e18b7c15414
spec:
  excludedNamespaces: null
  excludedResources:
  - apiservices
  - persistentvolumes
  - resticrepositories
  hooks:
    resources: null
  includeClusterResources: null
  includedNamespaces: null
  includedResources: null
  labelSelector:
    matchExpressions:
    - key: app
      operator: NotIn
      values:
      - ark
    - key: name
      operator: NotIn
      values:
      - restic
  snapshotVolumes: false
  storageLocation: default
  ttl: 168h0m0s
  volumeSnapshotLocations:
  - default
status:
  completionTimestamp: "2019-02-22T19:41:45Z"
  expiration: "2019-03-01T19:38:05Z"
  phase: Completed
  startTimestamp: "2019-02-22T19:38:05Z"
  validationErrors: null
  version: 1
  volumeSnapshotsAttempted: 0
  volumeSnapshotsCompleted: 0
```



* `ark restore describe <restorename>` or `kubectl get restore/<restorename> -n heptio-ark -o yaml`
```
Name:         totality
Namespace:    heptio-ark
Labels:       <none>
Annotations:  <none>

Backup:  cortex-ark-total-daily-20190222193143

Namespaces:
  Included:  *
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.ark.heptio.com, restores.ark.heptio.com
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto

Phase:  InProgress

Validation errors:  <none>

Warnings:  <none>
Errors:    <none>

Restic Restores (specify --details for more information):
  Completed:  1
```
* `ark restore logs <restorename>`
```
An error occurred: unable to retrieve logs because restore is not complete
```


**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Velero version (use `ark version`):  *v0.10.1*
- Kubernetes version (use `kubectl version`): *v1.12.5*
- Kubernetes installer & version: *Azure AKS*
- Cloud provider or hardware configuration: *Microsoft Azure*
- OS (e.g. from `/etc/os-release`): Unknown -- Azure AKS uses ubuntu images, I think?

--- a/changelogs/unreleased/1235-skriss
+++ b/changelogs/unreleased/1235-skriss
@@ -1,0 +1,1 @@
+Fix concurrency bug in code ensuring restic repository exists

--- a/pkg/restic/repository_ensurer.go
+++ b/pkg/restic/repository_ensurer.go
@@ -35,18 +35,31 @@ import (
 
 // repositoryEnsurer ensures that Velero restic repositories are created and ready.
 type repositoryEnsurer struct {
+	log        logrus.FieldLogger
 	repoLister velerov1listers.ResticRepositoryLister
 	repoClient velerov1client.ResticRepositoriesGetter
 
 	readyChansLock sync.Mutex
 	readyChans     map[string]chan *velerov1api.ResticRepository
+
+	// repoLocksMu synchronizes reads/writes to the repoLocks map itself
+	// since maps are not threadsafe.
+	repoLocksMu sync.Mutex
+	repoLocks   map[repoKey]*sync.Mutex
+}
+
+type repoKey struct {
+	volumeNamespace string
+	backupLocation  string
 }
 
 func newRepositoryEnsurer(repoInformer velerov1informers.ResticRepositoryInformer, repoClient velerov1client.ResticRepositoriesGetter, log logrus.FieldLogger) *repositoryEnsurer {
 	r := &repositoryEnsurer{
+		log:        log,
 		repoLister: repoInformer.Lister(),
 		repoClient: repoClient,
 		readyChans: make(map[string]chan *velerov1api.ResticRepository),
+		repoLocks:  make(map[repoKey]*sync.Mutex),
 	}
 
 	repoInformer.Informer().AddEventHandler(
@@ -67,7 +80,7 @@ func newRepositoryEnsurer(repoInformer velerov1informers.ResticRepositoryInforme
 					}
 
 					readyChan <- newObj
-					delete(r.readyChans, newObj.Name)
+					delete(r.readyChans, key)
 				}
 			},
 		},
@@ -84,6 +97,30 @@ func repoLabels(volumeNamespace, backupLocation string) labels.Set {
 }
 
 func (r *repositoryEnsurer) EnsureRepo(ctx context.Context, namespace, volumeNamespace, backupLocation string) (*velerov1api.ResticRepository, error) {
+	log := r.log.WithField("volumeNamespace", volumeNamespace).WithField("backupLocation", backupLocation)
+
+	// It's only safe to have one instance of this method executing concurrently for a
+	// given volumeNamespace + backupLocation, so synchronize based on that. It's fine
+	// to run concurrently for *different* namespaces/locations. If you had 2 goroutines
+	// running this for the same inputs, both might find no ResticRepository exists, then
+	// both would create new ones for the same namespace/location.
+	//
+	// This issue could probably be avoided if we had a deterministic name for
+	// each restic repository, and we just tried to create it, checked for an
+	// AlreadyExists err, and then waited for it to be ready. However, there are
+	// already repositories in the wild with non-deterministic names (i.e. using
+	// GenerateName) which poses a backwards compatibility problem.
+	log.Debug("Acquiring lock")
+
+	repoMu := r.repoLock(volumeNamespace, backupLocation)
+	repoMu.Lock()
+	defer func() {
+		repoMu.Unlock()
+		log.Debug("Released lock")
+	}()
+
+	log.Debug("Acquired lock")
+
 	selector := labels.SelectorFromSet(repoLabels(volumeNamespace, backupLocation))
 
 	repos, err := r.repoLister.ResticRepositories(namespace).List(selector)
@@ -97,11 +134,14 @@ func (r *repositoryEnsurer) EnsureRepo(ctx context.Context, namespace, volumeNam
 		if repos[0].Status.Phase != velerov1api.ResticRepositoryPhaseReady {
 			return nil, errors.New("restic repository is not ready")
 		}
+
+		log.Debug("Ready repository found")
 		return repos[0], nil
 	}
 
-	// no repo found: create one and wait for it to be ready
+	log.Debug("No repository found, creating one")
 
+	// no repo found: create one and wait for it to be ready
 	repo := &velerov1api.ResticRepository{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace:    namespace,
@@ -136,4 +176,20 @@ func (r *repositoryEnsurer) getReadyChan(name string) chan *velerov1api.ResticRe
 
 	r.readyChans[name] = make(chan *velerov1api.ResticRepository)
 	return r.readyChans[name]
+}
+
+func (r *repositoryEnsurer) repoLock(volumeNamespace, backupLocation string) *sync.Mutex {
+	r.repoLocksMu.Lock()
+	defer r.repoLocksMu.Unlock()
+
+	key := repoKey{
+		volumeNamespace: volumeNamespace,
+		backupLocation:  backupLocation,
+	}
+
+	if r.repoLocks[key] == nil {
+		r.repoLocks[key] = new(sync.Mutex)
+	}
+
+	return r.repoLocks[key]
 }

